rfcs/text/0010_service_status.md

```diff
@@ -0,0 +1,337 @@
+  meta?: JSONObject;
+} | {
+  level: ServiceStatusLevel;
+  summary: string; // required when level !== available
+  detail?: string;
+  documentationUrl?: string;
+  meta?: object;
+}
+
+/**
+ * The current "level" of availability of a service.
+ */
+enum ServiceStatusLevel {
+  /**
+   * Everything is working!
+   */
+  available,
+  /**
+   * Some features may not be working.
+   */
+  degraded,
+  /**
+   * The service is unavailable, but other functions that do not depend on this service should work.
+   */
+  unavailable,
+  /**
+   * Block all user functions and display the status page, reserved for Core services only.
+   * Note: In the real implementation, this will be split out to a different type. Kept as a single type here to make
+   * the RFC easier to follow.
+   */
+  critical
+}
+
+/**
+ * Status of core services. Only contains entries for backend services that could have a non-available `status`.
+ * For example, `context` cannot possibly be broken, so it is not included.
+ */
+interface CoreStatus {
+  elasticsearch: ServiceStatus;
+  http: ServiceStatus;
+  savedObjects: ServiceStatus;
+  uiSettings: ServiceStatus;
+  metrics: ServiceStatus;
+}
+
+// Types that specify valid JSON values.
+type JSONValue = string | number | boolean | null | JSONValue[] | JSONObject;
+type JSONObject = Record<string, JSONValue>;
+```
+
+## API Design
+
+### Plugin API
+
+```ts
+/**
+ * The API exposed to plugins on CoreSetup.status
+ */
+interface StatusSetup {
+  /**
+   * Allows a plugin to specify a custom status dependent on its own criteria.
+   * Completely overrides the default inherited status.
+   */
+  set(status$: Observable<ServiceStatus>): void;
+
+  /**
+   * Current status for all Core services.
+   */
+  core$: Observable<CoreStatus>;
+
+  /**
+   * Current status for all dependencies of the current plugin.
+   * Each key of the `Record` is a plugin id.
+   */
+  plugins$: Observable<Record<string, ServiceStatus>>;
+
+  /**
+   * The default status of the plugin as specified in the "Status inheritance"
+   * section below.
+   * 
+   * Useful when overriding the defaults with `set`.
+   */
+  inherited$: Observable<ServiceStatus>;
+}
+```
+
+### HTTP API
+
+The HTTP endpoint should return basic information about the Kibana node as well as the overall system status and the status of each individual system.
+
+This API does not need to include UI-specific details like the existing API such as `uiColor` and `icon`.
+
+```ts
+/**
+ * Response type for the endpoint: GET /api/status
+ */
+interface StatusResponse {
+  /** server.name */
+  name: string;
+  /** server.uuid */
+  uuid: string;
+  /** Currently exposed by existing status API */
+  version: {
+    number: string;
+    build_hash: string;
+    build_number: number;
+    build_snapshot: boolean;
+  };
+  /** Similar format to existing API, but slightly different shape */
+  status: {
+    /** See "Overall status calculation" section below */
+    overall: ServiceStatus;
```

Comment:
should this also return the `summary` string?

Reply:
Yes, `ServiceStatus` is the entire status interface, not just the level.